Content.Server/Preferences/PreferencesDatabase.cs

```diff
@@ -0,0 +1,210 @@
+                        {
+                            HairStyleName = reader.GetString(4),
+                            HairColor = Color.FromHex(reader.GetString(5)),
+                            FacialHairStyleName = reader.GetString(6),
+                            FacialHairColor = Color.FromHex(reader.GetString(7)),
+                            EyeColor = Color.FromHex(reader.GetString(8)),
+                            SkinColor = Color.FromHex(reader.GetString(9))
+                        }
+                    };
+                }
+
+                return new PlayerPreferences
+                {
+                    SelectedCharacterIndex = prefs.SelectedCharacterIndex,
+                    Characters = profiles.ToList()
+                };
+            }
+        }
+
+        private const string SaveSelectedCharacterIndexQuery =
+            @"UPDATE PlayerPreferences
+              SET SelectedCharacterIndex = @SelectedCharacterIndex
+              WHERE Username = @Username";
+
+        public void SaveSelectedCharacterIndex(string username, int index)
+        {
+            using (var connection = GetDbConnection())
+            {
+                connection.Execute(SaveSelectedCharacterIndexQuery,
+                    new {SelectedCharacterIndex = index, Username = username});
+            }
+        }
+
+        // https://stackoverflow.com/a/38463024
+        private const string SaveCharacterSlotQuery =
+            @"UPDATE HumanoidCharacterProfiles
+              SET
+              Name = @Name,
+              Age = @Age,
+              Sex = @Sex,
+              HairStyleName = @HairStyleName,
+              HairColor = @HairColor,
+              FacialHairStyleName = @FacialHairStyleName,
+              FacialHairColor = @FacialHairColor,
+              EyeColor = @EyeColor,
+              SkinColor = @SkinColor
+              WHERE Slot = @Slot AND Player = (SELECT Id FROM PlayerPreferences WHERE Username = @Username);
+              
+              -- If no update happened (i.e. the row didn't exist) then insert one
+              INSERT INTO HumanoidCharacterProfiles
+              (Name,
+              Age,
+              Sex,
+              HairStyleName,
+              HairColor,
+              FacialHairStyleName,
+              FacialHairColor,
+              EyeColor,
+              SkinColor)
+              VALUES
+              (@Name,
+              @Age,
+              @Sex,
+              @HairStyleName,
+              @HairColor,
+              @FacialHairStyleName,
+              @FacialHairColor,
+              @EyeColor,
+              @SkinColor)
+              WHERE (SELECT Changes() = 0);";
+
+        public void SaveCharacterSlot(string username, ICharacterProfile profile, int slot)
+        {
+            if (!(profile is HumanoidCharacterProfile humanoid))
+            {
+                // TODO: Handle other ICharacterProfile implementations properly
```

Comment:
How about:
```csharp
public interface ICharacterProfile {
    ...
    public (string query, object params) SaveProfileSQL() {
        ...
    }
    ...
}
```

the power of OOP

Reply:
Yeah, this is mostly just thrown together. Ideally we'd leverage ExposeData instead, I think?

Reply:
Depends how addicted to yaml the ExposeData system is guess.

Reply:
In theory not, but due to the design you can't really use it for something like SQL. The serializer doesn't know about field specification ahead of time so you couldn't form an SQL query from it.